### PR TITLE
Add support for Platform.isMacCatalyst

### PR DESCRIPTION
**Pull request by jacobp100**
## Summary:

Add support for `Platform.isMacCatalyst`

By default, Mac catalyst reports the idiom as an iPad, but you can check if it is Mac catalyst with a macro

There is technically the possibility to have the idiom return as a Mac, but that's not the default and almost definitely doesn't work in RN

## Changelog:

[IOS] [ADDED] Add support for `Platform.isMacCatalyst`

## Test Plan:

It compiles

### CODE REVIEW

File: packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ export interface Spec extends TurboModule {
     osVersion: string,
     systemName: string,
     interfaceIdiom: string,
+    isMacCatalyst: boolean,
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js:28]: I'm not sure this approach is scalable.. 
we already have iOS, iPadOS and Catalyst. Soon, we will also have something for the xrOS in compatibility mode. 🤔 

What if we call this `applePlatform` and we return a string enum with `'iPhone' | 'iPad' | 'Catalyst' | 'xrOS' | 'unknown' as values?

**Review by cipolleschi — CHANGES_REQUESTED** [review]
The code for the feature seems good to me. I have some concerns on the scalability of the solution, so I suggested something different.

Let me know what do you think!

File: packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js (hunk 1 above)
Comment (jacobp100) [reply packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js:28]: The difficulty is the idiom is either an iPad, or a Mac on Mac Catalyst - you get to pick

Comment (cipolleschi) [reply packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js:28]: I understand, but I don't get how that is related with the suggested change. Could you expand on that?
I'm suggsting a different way to represent the same information: we can populate the `applePlatform` variable with a similar logic, by using various `TARGET_OS_XXX` macros and then you can handle it in any way you want in JS, no?

**Comment by jacobp100** [comment]
Let me know if you want to do with the `idiom: "Catalyst-iPad" | "Catalyst-Mac"` or keep as-is

**Comment by cipolleschi** [comment]
can you rebase and fix conflicts?

**Comment by jacobp100** [comment]
@cipolleschi done

File: packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ export interface Spec extends TurboModule {
...
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/NativePlatformConstantsIOS.js:29]:
```suggestion
    isMacCatalyst?: boolean,
```

File: packages/react-native/Libraries/Utilities/Platform.d.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ type PlatformConstants = {
 };
 interface PlatformStatic {
   isTV: boolean;
+  isMacCatalyst: boolean;
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/Platform.d.ts:32]:
```suggestion
  isMacCatalyst?: boolean | undefined;
```

File: packages/react-native/Libraries/Utilities/Platform.d.ts [hunk 4]
```diff
@@ -54,6 +55,7 @@ interface PlatformIOSStatic extends PlatformStatic {
   OS: 'ios';
   isPad: boolean;
   isTV: boolean;
+  isMacCatalyst: boolean;
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/Platform.d.ts:58]:
```suggestion
  isMacCatalyst?: boolean | undefined;
```

File: packages/react-native/Libraries/Utilities/Platform.ios.js [hunk 5]
```diff
@@ -39,6 +39,7 @@ const Platform = {
       prerelease: ?number,
     |},
     systemName: string,
+    isMacCatalyst: boolean,
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/Platform.ios.js:42]:
```suggestion
    isMacCatalyst?: boolean,
```

File: packages/react-native/Libraries/Utilities/Platform.ios.js [hunk 6]
```diff
@@ -71,6 +72,11 @@ const Platform = {
     // $FlowFixMe[object-this-reference]
     return this.constants.isDisableAnimations ?? this.isTesting;
   },
+  // $FlowFixMe[unsafe-getters-setters]
+  get isMacCatalyst(): boolean {
+    // $FlowFixMe[object-this-reference]
+    return this.constants.isMacCatalyst;
```
Comment (cipolleschi) [inline packages/react-native/Libraries/Utilities/Platform.ios.js:78]:
```suggestion
    return this.constants.isMacCatalyst ?: false;
```

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @jacobp100 I tried to land this internally, but we have some issues as the JS-side update will break our infra because it is not backward compatible.

I added some suggestions to make the field optional for the time being so it will work correctly and in a few weeks we can remove the optionality. How does this sound?

**Comment by cipolleschi** [comment]
@jacobp100, I see that the CI is red. Could you have a look at it and fix the issues, please?  🙏

**Comment by jacobp100** [comment]
@cipolleschi sorry don't know how that happened. All green now

**Comment by jacobp100** [comment]
@cipolleschi I've brought this PR up-to-date with main again. Sorry for the number of commits - I did it all via the GH website

**Review by cipolleschi — CHANGES_REQUESTED** [review]
We need to make this property optional, otherwise our compatibility checks will fail.

**Pull request closed without merging**